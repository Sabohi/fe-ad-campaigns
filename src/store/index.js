import { routerMiddleware } from 'connected-react-router';
import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import history from '@/routes/History';
import rootReducer from './root-reducer';

const middlewares = [thunk, routerMiddleware(history)];

const persistConfig = {
  storage,
  key: 'root',
};

const composeEnhancers = (process.env.NODE_ENV === 'development' && window !== undefined && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const persistedReducer = persistReducer(persistConfig, rootReducer(history));

const store = createStore(persistedReducer, composeEnhancers(applyMiddleware(...middlewares)));

const persistor = persistStore(store);

export { store, persistor };
