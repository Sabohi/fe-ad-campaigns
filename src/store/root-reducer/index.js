import { combineReducers } from 'redux';
import { connectRouter } from 'connected-react-router';
import userReducer from '@/redux/users/UserReducer';

export default (history) =>
  combineReducers({
    router: connectRouter(history),
    userReducer,
  });
