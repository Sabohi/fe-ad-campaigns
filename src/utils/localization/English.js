export default {
  header: {
    projectHeader: 'Ad Campaigns',
  },
  common: {
    noDataFound: 'No Data Found',
    unknownUser: 'Unknown User',
    active: 'Active',
    inactive: 'Inactive',
    currency: 'USD',
    na: 'NA',
  },
  placeHolders: {
    search: 'Search by Name',
  },
  tableColumns: {
    name: 'Name',
    userName: 'User Name',
    startDate: 'Start Date',
    endDate: 'End Date',
    status: 'Active',
    budget: 'Budget',
    dateShow: 'Date Range',
  },
};
