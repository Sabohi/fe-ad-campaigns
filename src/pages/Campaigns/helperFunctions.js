import localization from '@/utils/localization';
import { convertToICS } from '@/utils/commonFunctions/CommonFunctions';
import moment from 'moment';
import { GLOBALCONSTANTS } from '@/utils/GlobalConstants';

export const prepareData = (receivedData, userList) => {
  const data = [...receivedData.values()];
  const dateFormat = GLOBALCONSTANTS.DATE_FORMAT;
  console.warn = jest.fn();
  // Filter out data which contains startDate less than end date but rws with blank start or end will be shown.
  const dataFilter = data.length
    ? data.filter((item) => {
        let ret = typeof item?.id === 'number';
        if (item?.startDate && item.endDate) {
          ret = moment(item.startDate).isSameOrBefore(item.endDate);
        }
        return ret;
      })
    : [];
  const resultData = dataFilter.length
    ? dataFilter.map((item) => {
        let userName = localization.common.unknownUser;
        if (item?.userId) {
          const findItem = userList.findIndex((users) => users.id === item.userId);
          if (findItem !== -1) {
            userName = userList[findItem].name;
          }
        }
        const startDate = moment(item.startDate).format(dateFormat);
        const endDate = moment(item.endDate).format(dateFormat);

        // For those campaigns whose start or end is not available, status is shown as inactive
        const status = moment().isBetween(startDate, endDate);

        return [item.name, userName, startDate || localization.common.na, endDate || localization.common.na, status, convertToICS(item.Budget)];
      })
    : [];
  return resultData;
};
