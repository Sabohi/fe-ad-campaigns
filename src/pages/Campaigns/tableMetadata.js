import React from 'react';
import FormControlLabel from '@mui/material/FormControlLabel';
import FormGroup from '@mui/material/FormGroup';
import FormLabel from '@mui/material/FormLabel';
import moment from 'moment';
import MomentUtils from '@date-io/moment';
import { DatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import localization from '@/utils/localization';
import { GLOBALCONSTANTS } from '@/utils/GlobalConstants';

const dateFormat = GLOBALCONSTANTS.DATE_FORMAT;

export const columns = [
  {
    name: localization.tableColumns.name,
    options: { filter: false },
  },
  {
    name: localization.tableColumns.userName,
    options: { filter: false, searchable: false },
  },
  {
    name: localization.tableColumns.startDate,
    options: {
      filter: false,
      searchable: false,
    },
  },
  {
    name: localization.tableColumns.endDate,
    options: {
      filter: false,
      searchable: false,
    },
  },
  {
    name: localization.tableColumns.status,
    options: {
      filter: false,
      sort: false,
      searchable: false,
      customBodyRender: (value) => (
        <FormControlLabel
          control={<span className={`status ${value ? 'active' : 'inactive'}`} data-testid="campaign-row" />}
          label={value ? localization.common.active : localization.common.inactive}
        />
      ),
    },
  },
  {
    name: localization.tableColumns.budget,
    options: { filter: false, searchable: false },
  },
  {
    name: localization.tableColumns.dateShow,
    options: {
      display: false,
      filter: true,
      filterType: 'custom',
      filterList: [],
      customFilterListOptions: {
        render: (v) => {
          if (v[0] && v[1]) {
            return [`${localization.tableColumns.startDate}: ${v[0]}`, `${localization.tableColumns.endDate}: ${v[1]}`];
          }
          if (v[0]) {
            return `${localization.tableColumns.startDate}: ${v[0]}`;
          }
          if (v[1]) {
            return `${localization.tableColumns.endDate}: ${v[1]}`;
          }
          return false;
        },
        update: (filterList, filterPos, index) => {
          if (filterPos === 0) {
            filterList[index].splice(filterPos, 1, null);
          } else if (filterPos === 1) {
            filterList[index].splice(filterPos, 1);
          } else if (filterPos === -1) {
            filterList[index] = [];
          }
          return filterList;
        },
      },
      filterOptions: {
        names: [],
        logic(val, filters, row) {
          const startCheck = filters[0];
          const endCheck = filters[1];
          let startValid = false;
          let endValid = false;
          if (startCheck) {
            const checkDate = row[2];
            const startCom = moment(checkDate).isSameOrAfter(filters[0]);
            startValid = !startCom;
          }
          if (endCheck) {
            const checkDate = row[3];
            const endCom = moment(checkDate).isSameOrBefore(filters[1]);
            endValid = !endCom;
          }
          return startValid || endValid;
        },
        display: (filterList, onChange, index, column) => {
          const momentDate = moment();
          const minEndDate = filterList[index][0] || momentDate.format(dateFormat);
          const checkYear = momentDate.format('YY');
          const maxStartDate = filterList[index][1] || momentDate.add(100 - checkYear, 'years').format(dateFormat);

          return (
            <>
              <FormLabel>Date</FormLabel>
              <FormGroup row>
                <MuiPickersUtilsProvider libInstance={moment} utils={MomentUtils}>
                  <DatePicker
                    autoOk
                    label={localization.tableColumns.startDate}
                    value={filterList[index][0] || null}
                    format={dateFormat}
                    clearable
                    maxDate={maxStartDate}
                    onChange={(event) => {
                      const updateVal = event ? event.format(dateFormat) : null;
                      if (updateVal) {
                        filterList[index][0] = updateVal;
                      } else {
                        filterList[index] = [];
                      }
                      onChange(filterList[index], index, column);
                    }}
                  />
                  <DatePicker
                    autoOk
                    label={localization.tableColumns.endDate}
                    value={filterList[index][1] || null}
                    format={dateFormat}
                    clearable
                    disabled={!filterList[index][0]}
                    minDate={minEndDate}
                    onChange={(event) => {
                      const updateVal = event ? event.format(dateFormat) : null;
                      if (!updateVal && !filterList[index][0]) {
                        filterList[index] = [];
                      } else {
                        filterList[index][1] = updateVal;
                      }
                      onChange(filterList[index], index, column);
                    }}
                  />
                </MuiPickersUtilsProvider>
              </FormGroup>
            </>
          );
        },
      },
      print: false,
      searchable: false,
    },
  },
];

export const options = {
  selectableRows: 'none',
  search: true,
  searchPlaceholder: localization.placeHolders.search,
  download: false,
  print: false,
  viewColumns: false,
  filter: true,
  filterType: 'dropdown',
  responsive: 'vertical',
  tableBodyHeight: 'calc(100vh - 6.5rem)',
  tableBodyMaxHeight: 'calc(100vh - 6.5rem)',
  textLabels: {
    body: {
      noMatch: localization.common.noDataFound,
    },
  },
};
