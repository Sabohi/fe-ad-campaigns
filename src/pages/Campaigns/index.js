import DataTable from '@/components/DataTable';
import React, { useState, useEffect, useCallback } from 'react';
import { getUsersInfo } from '@/redux/users/UserAction';
import localization from '@/utils/localization';
import { useDispatch, useSelector } from 'react-redux';
import useLoader from '@/hooks/useLoader';
import { prepareData } from './helperFunctions';
import { columns, options } from './tableMetadata';
import { initalCampaignData } from './initalCampaignData';

function CampaignList() {
  const dispatch = useDispatch();
  const userList = useSelector((state) => state.userReducer.userList);
  const [mounted, setMounted] = useState(false);
  const [data, setData] = useState(initalCampaignData);
  const [inputData, setInputData] = useState(new Map());
  const [loader, startLoader, stopLoader] = useLoader();

  if (!mounted) {
    dispatch(getUsersInfo());
  }

  useEffect(() => {
    updateInputData(initalCampaignData);
    setMounted(true);
  }, [updateInputData]);

  useEffect(() => {
    const processdData = prepareData(inputData, userList);
    setData(processdData);
  }, [userList, inputData]);

  const updateInputData = useCallback(
    (receivedData = []) => {
      const checkArr = Array.isArray(receivedData);
      if (checkArr) {
        startLoader();
        const processedData = new Map();
        receivedData.forEach((item) => {
          processedData.set(item?.id, item);
        });
        const updatedInput = new Map([...inputData, ...processedData]);
        setTimeout(() => {
          setInputData(updatedInput);
          stopLoader();
        }, 3000);
      }
    },
    [inputData, startLoader, stopLoader],
  );

  window.AddCampaigns = updateInputData;

  return (
    <>
      {loader}
      <DataTable title={localization.header.campaignHeading} data={data} columns={columns} options={options} />
    </>
  );
}

export default CampaignList;
