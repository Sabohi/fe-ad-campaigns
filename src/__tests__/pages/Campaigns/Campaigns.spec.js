import { cleanup, render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { StaticRouter } from 'react-router';
import { store } from '@/store';
import Campaigns from '@/pages/Campaigns';
import { shallow } from 'enzyme';
import localization from '@/utils/localization';

describe('Testing Campaign module mounting', () => {
  it('Campaigns snapshot test', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <Campaigns />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Campaigns module testing', () => {
  let wrapper;
  beforeEach(() => {
    cleanup();
    wrapper = render(
      <Provider store={store}>
        <StaticRouter>
          <Campaigns />
        </StaticRouter>
      </Provider>,
    );
  });
  it('it checks "No Data Found" case', () => {
    expect(wrapper.getByText(/no data/i).textContent).toBe(localization.common.noDataFound);
  });
  it('it checks that table is in document', () => {
    expect(wrapper.getByLabelText('Table Toolbar')).toBeInTheDocument();
  });
  it('it checks that search bar is mounted', () => {
    expect(wrapper.getByTestId('Search-iconButton')).toBeInTheDocument();
  });
  it('it checks that filter table is mounted', () => {
    expect(wrapper.getByTestId('Filter Table-iconButton')).toBeInTheDocument();
  });
});
