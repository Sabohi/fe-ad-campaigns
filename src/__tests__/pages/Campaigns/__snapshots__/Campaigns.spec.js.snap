// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ad campaign module mounting Campaigns snapshot test 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <CampaignList />
</ContextProvider>
`;
